datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt

```diff
@@ -56,57 +59,129 @@ internal object ResourceMapper {
+      extractField(
+        questionnaireItemListIterator.next(),
+        questionnaireResponseItemListIterator.next()
+      )
+    }
+  }
+
+  /**
+   * Extracts the answer value from [questionnaireResponseItem] and updates the field defined in
+   * [questionnaireItem]. This method handles nested fields.
+   */
+  private fun Base.extractField(
+    questionnaireItem: Questionnaire.QuestionnaireItemComponent,
+    questionnaireResponseItem: QuestionnaireResponse.QuestionnaireResponseItemComponent
+  ) {
+    if (questionnaireItem.definition == null) {
+      extractFields(questionnaireItem.item, questionnaireResponseItem.item)
+      return
+    }
+
+    val targetFieldName = questionnaireItem.definitionFieldName ?: return
+    if (targetFieldName.isEmpty()) {
+      return
+    }
+
+    val definitionField = questionnaireItem.getDefinitionField ?: return
+    if (questionnaireItem.type == Questionnaire.QuestionnaireItemType.GROUP) {
+      val value: Base =
+        (definitionField.nonParameterizedType.newInstance() as Base).apply {
+          extractFields(questionnaireItem.item, questionnaireResponseItem.item)
+        }
+
+      updateField(definitionField, value)
+    } else {
+      if (questionnaireResponseItem.answer.isEmpty()) return
+      val answer = questionnaireResponseItem.answer.first().value
+
+      if (!definitionField.nonParameterizedType.isEnum) {
+        // this is a low level type e.g. StringType
+        updateField(definitionField, answer)
+      } else {
+        // this is a high level type e.g. AdministrativeGender
+        updateFieldWithEnum(definitionField, targetFieldName, answer)
+      }
+    }
   }
 }
 
 /**
- * Extracts values for fields in the builder from the corresponding questions and answers in
- * [questionnaireItemList] and [questionnaireResponseItemList].
+ * Updates a field of name [targetFieldName] on this object with the generated enum from [answer]
+ * using the declared setter. [propertyType] helps to determine the field class type. The enum is
+ * generated by calling fromCode method on the enum class
  */
-private fun Base.extractFields(
-  questionnaireItemList: List<Questionnaire.QuestionnaireItemComponent>,
-  questionnaireResponseItemList: List<QuestionnaireResponse.QuestionnaireResponseItemComponent>
-) {
-  val questionnaireItemListIterator = questionnaireItemList.iterator()
-  val questionnaireResponseItemListIterator = questionnaireResponseItemList.iterator()
-  while (questionnaireItemListIterator.hasNext() &&
-    questionnaireResponseItemListIterator.hasNext()) {
-    val questionnaireItem = questionnaireItemListIterator.next()
-    val questionnaireResponseItem = questionnaireResponseItemListIterator.next()
-    this.extractField(questionnaireItem, questionnaireResponseItem)
-    extractFields(questionnaireItem.item, questionnaireResponseItem.item)
-  }
+private fun Base.updateFieldWithEnum(propertyType: Field, targetFieldName: String, answer: Base) {
```

Comment:
similar to `updateField` method, the `targetFieldName` can be removed here no?

also rename `propertyType` to `field` and `answer` to `value`.

```suggestion
private fun Base.updateFieldWithEnum(field: Field, value: Base) {
```

Reply:
Fixed